core/web/loop_registry.go

```diff
@@ -0,0 +1,119 @@
+package web
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"sort"
+	"strings"
+
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+
+	"github.com/smartcontractkit/chainlink/v2/core/services/chainlink"
+	"github.com/smartcontractkit/chainlink/v2/plugins"
+)
+
+const invalidPort = -1
+
+type LoopRegistry struct {
+	exposedPromPort int
+	// read only so no mutex needed
+	pluginLookupFn func() map[string]plugins.EnvConfigurer
+}
+
+func NewLoopRegistry(app chainlink.Application) *LoopRegistry {
+	return &LoopRegistry{
+		exposedPromPort: int(app.GetConfig().Port()),
+		pluginLookupFn:  app.GetLoopEnvConfig,
+	}
+}
+
+type pluginConfig struct {
+	name string
+	port int
+}
+
+// list returns deterministic list of loop's known the registry
+func (l *LoopRegistry) list() []pluginConfig {
+	var out []pluginConfig
+	for name, cfg := range l.pluginLookupFn() {
+		out = append(out, pluginConfig{name: name, port: cfg.PrometheusPort()})
+	}
+	sort.Slice(out, func(i, j int) bool {
+		return out[i].port < out[j].port
+	})
+	return out
+}
+
+func (l *LoopRegistry) get(name string) (pluginConfig, bool) {
+	result := pluginConfig{name: name, port: invalidPort}
+	envCfg, exists := l.pluginLookupFn()[name]
+	if exists {
+		result.port = envCfg.PrometheusPort()
+	}
+	return result, exists
+}
+
+func (l *LoopRegistry) discoveryHandler(w http.ResponseWriter, req *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+	groups := make([]*targetgroup.Group, 0)
```

Comment:
Maybe simplify? Or include cap?
```suggestion
	var groups []*targetgroup.Group
```